File: src/VoidRay/Services/Format.cs
namespace VoidRay.Services;

public static class Format
{
    private static readonly string[] UnitsEn = { "B", "KB", "MB", "GB", "TB", "PB" };
    private static readonly string[] UnitsFr = { "o", "Ko", "Mo", "Go", "To", "Po" };

    /// <summary>Human readable size, e.g. "8.92 GB" / "8,92 Go".</summary>
    public static string Bytes(double bytes)
    {
        var units = Loc.I.Lang == "fr" ? UnitsFr : UnitsEn;
        var value = Math.Max(0, bytes);
        var unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }
        var format = unit == 0 || value >= 100 ? "0" : "0.##";
        return value.ToString(format, Loc.I.Culture) + " " + units[unit];
    }

    public static string ShortDate(DateTimeOffset date) =>
        date.ToLocalTime().ToString("d MMM yyyy", Loc.I.Culture);

    public static string Stamp(DateTimeOffset date) =>
        ShortDate(date) + " · " + date.ToLocalTime().ToString("HH:mm", Loc.I.Culture);

    public static string Ago(DateTimeOffset date)
    {
        var diff = DateTimeOffset.Now - date;
        if (diff < TimeSpan.FromMinutes(2)) return Loc.T("justNow");
        if (diff < TimeSpan.FromHours(1)) return Loc.T("minsAgo", (int)Math.Round(diff.TotalMinutes));
        if (diff < TimeSpan.FromDays(1)) return Loc.T("hoursAgo", (int)Math.Round(diff.TotalHours));
        return Loc.T("daysAgo", (int)Math.Round(diff.TotalDays));
    }

    public static string Duration(TimeSpan t) =>
        t.TotalHours >= 1 ? $"{(int)t.TotalHours:00}:{t.Minutes:00}:{t.Seconds:00}" : $"{t.Minutes:00}:{t.Seconds:00}";
}

File: android/app/src/main/java/win/enderr/voidray/core/Format.kt
package win.enderr.voidray.core

import win.enderr.voidray.i18n.L
import java.text.DecimalFormat
import java.text.DecimalFormatSymbols
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlin.math.roundToInt

object Format {
    private val unitsEn = listOf("B", "KB", "MB", "GB", "TB", "PB")
    private val unitsFr = listOf("o", "Ko", "Mo", "Go", "To", "Po")

    /** Human readable size, e.g. "8.92 GB" / "8,92 Go". */
    fun bytes(bytes: Double): String {
        val units = if (L.lang == "fr") unitsFr else unitsEn
        var value = maxOf(0.0, bytes)
        var unit = 0
        while (value >= 1024 && unit < units.size - 1) {
            value /= 1024; unit++
        }
        val pattern = if (unit == 0 || value >= 100) "0" else "0.##"
        return DecimalFormat(pattern, DecimalFormatSymbols(L.locale)).format(value) + " " + units[unit]
    }

    fun bytes(bytes: Long) = bytes(bytes.toDouble())

    private fun zoned(ms: Long) = Instant.ofEpochMilli(ms).atZone(ZoneId.systemDefault())

    fun shortDate(ms: Long): String = zoned(ms).format(DateTimeFormatter.ofPattern("d MMM yyyy", L.locale))

    fun stamp(ms: Long): String = shortDate(ms) + " · " + zoned(ms).format(DateTimeFormatter.ofPattern("HH:mm", L.locale))

    fun ago(ms: Long): String {
        val diff = System.currentTimeMillis() - ms
        return when {
            diff < 120_000 -> L.t("justNow")
            diff < 3_600_000 -> L.t("minsAgo", (diff / 60_000.0).roundToInt())
            diff < 86_400_000 -> L.t("hoursAgo", (diff / 3_600_000.0).roundToInt())
            else -> L.t("daysAgo", (diff / 86_400_000.0).roundToInt())
        }
    }

    fun duration(seconds: Long): String {
        val h = seconds / 3600
        val m = (seconds % 3600) / 60
        val s = seconds % 60
        return if (h > 0) "%02d:%02d:%02d".format(h, m, s) else "%02d:%02d".format(m, s)
    }
}
